flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GroupAggProcessFunction.scala

```diff
@@ -84,16 +109,38 @@ class GroupAggProcessFunction(
     }
 
     // Set aggregate result to the final output
-    i = 0
-    while (i < aggregates.length) {
-      val index = groupings.length + i
-      val accumulator = accumulators.getField(i).asInstanceOf[Accumulator]
-      aggregates(i).accumulate(accumulator, input.getField(aggFields(i)(0)))
-      output.setField(index, aggregates(i).getValue(accumulator))
-      i += 1
+    if (input.command == Command.Delete) {
+      i = 0
+      while (i < aggregates.length) {
+        val index = groupings.length + i
+        val accumulator = accumulators.getField(i).asInstanceOf[Accumulator]
+        aggregates(i).retract(accumulator, input.getField(aggFields(i)(0)))
+        output.setField(index, aggregates(i).getValue(accumulator))
+        i += 1
+      }
+    } else {
+      i = 0
+      while (i < aggregates.length) {
+        val index = groupings.length + i
+        val accumulator = accumulators.getField(i).asInstanceOf[Accumulator]
+        aggregates(i).accumulate(accumulator, input.getField(aggFields(i)(0)))
+        output.setField(index, aggregates(i).getValue(accumulator))
+        i += 1
+      }
     }
-    state.update(accumulators)
 
+    // if previous is not null, do retraction process
+    if (null != previous) {
```

Comment:
check against `generateRetraction`. The check can be optimized because `generateRetraction`is a `val` and hence `final`.